### PR TITLE
Support for TUM Visual-Inertial Dataset

### PR DESCRIPTION
**Pull request by matsuren**
Hi! Thank you for the wonderful project! 
I modified the source code to run on [TUM Visual-Inertial Dataset](https://vision.in.tum.de/data/datasets/visual-inertial-dataset) by `run_euroc_slam`.   You need to merge #165 first since images from TUM VI are captured by superwide fisheye cameras (`min_y > max_y` problems mentioned in #83 ). 

I have questions:
I'd like to add calibration files that I'm using for TUM VI. But, where should I add them? `run_euroc_slam` is used to run on TUM VI so inside `openvslam/example/euroc` is fine? or should I create an additional folder `openvslam/example/tum_vi`?
FYI, I used the following script to generate the calibration YAML files.
https://nbviewer.jupyter.org/gist/matsuren/6c00e1b6e006df9ef87c3271316cd5f9

Also, let me know if you want me to add some instruction about TUM VI in the documents.

### CODE REVIEW

**Comment by matsuren** [comment]
Hi, 
It's just a passing thought, but maybe it's better to utilize 16 bit images? I will check if histogram equalizations of 16 bit images before converting into 8 bit grayscale make difference in a few days. 
Thanks,

**Comment by matsuren** [comment]
Hi,
I checked the histogram of images by the following code:
```python
import cv2
import numpy as np
from matplotlib import pyplot as plt
filename = 'dataset-room1_512_16/mav0/cam0/data/1520530308199447626.png'
fig, ax = plt.subplots(1,2, figsize=(12,5))

img = cv2.imread(filename, cv2.IMREAD_UNCHANGED)
ax[0].hist(img.ravel(), 256)
ax[0].set_title('cv2.IMREAD_UNCHANGED')
img = cv2.imread(filename, cv2.IMREAD_GRAYSCALE)
ax[1].hist(img.ravel(),256,[0,255])
ax[1].set_title('cv2.IMREAD_GRAYSCALE')
```
hist

As you can see this figure, histogram equalization like `convert16to8gray` might help. But I guess this is out of focus of this pull request, so I won't deal with it here.
```python
def convert16to8gray(img):
    l = np.percentile(img, 5)
    h = np.percentile(img, 95)
    img = (img-l)/(h-l)*255
    return img.clip(0, 255).astype(np.uint8)
```
If you have any suggestions, please let me know.
Thanks,

**Comment by shinsumicco** [comment]
@matsuren 
Sorry for the late reply.

It would be grateful if you could insert histogram equalization process as you suggested.
Are the final equalization results different between the following two schemes?
1. loading an image as **16bit**, equalizing it, and finally converting it to **8bit** (as you suggested)
2. loading an image as **8bit** from the beginning, and equalizing it

I'd like you to make the modification in this PR.
The following steps would be helpful to us:
- implementing a function to equalize histogram (e.g. `equalize_histogram()`) at [`./src/openvslam/util/image_converter.cc`](https://github.com/xdspacelab/openvslam/blob/9e722cc9ea24ca5b25d5dbe2fef16ec417162f5c/src/openvslam/util/image_converter.cc#L44)
- append an option to select whether to use the equalization or not in `./example/run_euroc_slam.cc`

Thanks for your beneficial contribution!

**Comment by matsuren** [comment]
Hi, thank you for your suggestion.
My intention for histogram equalization is to prevent some information loss when you convert 16 bit image into 8 bit image. So I'm not sure if equalizing histogram for 8 bit image will help.

The figure bellow shows the histogram of the images after equalization:
- loading an image as 16bit, equalizing it, and finally converting it to 8bit (as you suggested)
- loading an image as 8bit from the beginning, and equalizing it
histogram

I modified the source code according to your comment.
If you have any questions, feel free to ask me. 

Thanks,

File: src/openvslam/util/image_converter.cc [hunk 1]
```diff
@@ -42,5 +42,23 @@ void convert_to_true_depth(cv::Mat& img, const double depthmap_factor) {
     img.convertTo(img, CV_32F, 1.0 / depthmap_factor);
 }
 
+void equalize_histogram(cv::Mat& img) {
+    assert(img.type() == CV_8UC1 || img.type() == CV_16UC1);
+    if (img.type() == CV_16UC1) {
+        std::vector<unsigned short> vec(img.begin<unsigned short>(), img.end<unsigned short>());
+        std::sort(vec.begin(), vec.end());
+        auto l = vec.at(static_cast<unsigned int>(0.05 * vec.size()));
+        auto h = vec.at(static_cast<unsigned int>(0.95 * vec.size()));
```
[the remark this answers is not in the document]
Comment (matsuren) [reply src/openvslam/util/image_converter.cc]: Sorry, I was careless!

**Comment by matsuren** [comment]
Thanks for your comment.

>Did you calibrate parameters using some of the calibration sequences, or, reconfigure the calibrated parameters provided by TUM?

I reconfigured the calibrated parameters provided by [TUM here](https://vision.in.tum.de/data/datasets/visual-inertial-dataset#geometric_calibration), specifically [pinhole (512x512)](https://vision.in.tum.de/_media/data/datasets/visual-inertial-dataset/pinhole-equi-512.zip).
I used the following python script for the conversion. 
[Link to jupyter notebook](https://nbviewer.jupyter.org/gist/matsuren/6c00e1b6e006df9ef87c3271316cd5f9/TUM_VI_OpenVSLAM_calib.ipynb)


Thanks,

**Comment by shinsumicco** [comment]
@matsuren 
Thanks for the information about the calibration parameters.
I've confirmed the behavior with some of the TUM VI sequences.
Your PR would be very helpful to me for implementation of visual-inertial tracking algorithm.
Thank you for your continued support!

**Pull request merged**